src/createAnimatedComponent.tsx

```diff
@@ -226,15 +237,59 @@ interface ComponentRef extends Component {
   getScrollableNode?: () => ComponentRef;
 }
 
-export interface InitialComponentProps extends Record<string, unknown> {
-  ref?: Ref<Component>;
+export interface InitialFunctionComponentProps extends Record<string, unknown> {
   collapsable?: boolean;
 }
 
-export default function createAnimatedComponent(
-  Component: ComponentType<InitialComponentProps>,
-  options?: Options<InitialComponentProps>
-): ComponentType<AnimatedComponentProps<InitialComponentProps>> {
+export interface InitialComponentProps extends InitialFunctionComponentProps {
+  ref?: Ref<Component>;
+}
+
+export type Intersection<T, U> = Pick<
+  T,
+  Extract<keyof T, keyof U> & Extract<keyof U, keyof T>
+>;
+
+export type Diff<T, U> = Pick<T, SetDifference<keyof T, keyof U>>;
+
+export type SetDifference<A, B> = A extends B ? never : A;
+
+export type ExtendProps<
+  Props,
+  ExtendWith,
+  I = Diff<Props, ExtendWith> &
+    (Intersection<Props, ExtendWith> | Intersection<ExtendWith, Props>)
+> = Pick<I, keyof I>;
+
+type PropsExtensions = {
+  onScroll?: ReturnType<typeof useAnimatedScrollHandler>;
+  style?: NestedArray<StyleProps>;
+};
+
+function createAnimatedComponent<P>(
+  Component: FunctionComponent<P & InitialFunctionComponentProps>,
+  options?: Options<P & InitialFunctionComponentProps>
+): FunctionComponent<
+  AnimatedComponentProps<
+    ExtendProps<P, PropsExtensions> & InitialFunctionComponentProps
+  >
+>;
+function createAnimatedComponent<P>(
+  Component: ComponentType<P & InitialComponentProps>,
```

Comment:
```suggestion
  Component: ComponentClass<P & InitialComponentProps>,
```

Reply:
We still have some errors with createAnimatedComponent, we should circle back to this PR and fix them